### ISSUE TITLE
agreements terminated because of "workload terminated"

### ISSUE THREAD

**Issue by linggao** [open-horizon/anax#90]
In the repetitive cancellation tests, there are a lot of agreement got canceled because of "workload terminated". Checking the syslog, found that the workload could not start because there are no enough ip in the ip pool for it.  
The related errors are:
Jan 20 13:20:36 vs3 snap[21575]: E0120 13:20:36.818865   21622 container.go:943] Error starting containers: API error (500): {"message":"could not find an available, non-overlapping IPv4 address pool among the defaults to assign to the network"}

The network could not be removed for previous workload containers because of this:
Jan 20 12:56:35 vs3 snap[21575]: I0120 12:56:35.406252   21622 container.go:1092] Freeing agreement net: d0ca332359a73f5c0637121cca28169a8173f17403dbc19726b47c6db723b484
Jan 20 12:56:35 vs3 snap[21575]: E0120 12:56:35.411960   21622 container.go:1101] Failure removing network: 6ca3d3c42a6cf25edb6896e0f805808d36a3e5160895d58363371b34a5f34954. Error: API error (500): {"message":"network d0ca332359a73f5c0637121cca28169a8173f17403dbc19726b47c6db723b484 has active endpoints"}
Jan 20 12:56:35 vs3 snap[21575]: I0120 12:56:35.442092   21622 container.go:1120] Removing iptables isolation rules for agreement d0ca332359a73f5c0637121cca28169a8173f17403dbc19726b47c6db723b484

Jan 20 12:56:02 vs3 snap[21575]: E0120 12:56:02.384070   21622 container.go:1070] Service pvp in agreement d0ca332359a73f5c0637121cca28169a8173f17403dbc19726b47c6db723b484 could not be removed. Error: API error (500): {"message":"Unable to remove filesystem for 11ab08e88c742e6de498b28c6e3fbd0486fe72c3b111200bf1dd16b886c6f11b: remove /var/lib/docker/containers/11ab08e88c742e6de498b28c6e3fbd0486fe72c3b111200bf1dd16b886c6f11b/shm: device or resource busy"}

**Comment by linggao** [comment]
docker version
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:58:26 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:58:26 2017
 OS/Arch:      linux/amd64
 Experimental: false

**Comment by linggao** [comment]
Other people have the same issue.  https://github.com/docker/docker/issues/23302

**Comment by linggao** [comment]
It is resource issue for docker.  I changed the cancellation interval to 15min on the 1-cpu vm, the problem went away.  And I also tested it on a 4-cpu vm with cancellation intercal 10 minutes, and did not see such problem.

**Comment by linggao** [comment]
It may not be related, but there are places we need to do more checking before proceeding because different threads are handling different events at the same time. For example.
1. need to check (in api.go)  if the agreement is already terminated when handling user request to cancel an agreement.
2. need to check (in governance.go or container.go) if the agreement is already terminated when handling "configure" event.
I will fix these places and retest.

**Comment by linggao** [comment]
Added code mentioned above and retested, still seeing failure when removing containers and releasing docker networks from time to time.  This caused number of networks increasing. Once it reached 32, there can be no longer new containers. Hence all the consequent agreement got "workload terminated" errors.

**Comment by linggao** [comment]
Meesage from Carl:  docker has been having this issue forever ... it's still a thing and looks like it actually got worse in 1.12 (and apparently hasn't been solved in 1.13 either) ... see this discussion:
https://github.com/docker/docker/issues/22260

**Comment by linggao** [comment]
Boot also got ""workload terminated" in the termination description in his tests. But it was due to a different reason: 
16:27:53.682165     127 container.go:943] Error starting containers: mkdir /root/.colonus/anaxRO/869af01d1bafd8ed305c337f9e15a2baa34382dc42dbeb639ca7ec9efa9feeff: file exists"
Here is why: in govenence.go there are 2 commands calling RecordReply function which will in turn triggers the torrent downloading and container setup. One is the whisper rplyack, the other is the blockchain writing.   When handling blockchain writing message, we check if AgreementAcceptedTime == 0. We did not check for it for the other case.  So for the agreement mentioned, the rplyack came after the blockchain write notification. So container have received 2 events to start the containers for the same agreements.
Solution: we can add checking for AgreementAcceptedTime == 0 for the RplyAck case for now.  But it still cannot guarantee when these 2 messages come very close that both threads have AgreementAcceptedTime  equal to 0 at the time when the commend is processed.  So we need to make sure our torrent fetching and container setup functions can tolerant multiple entrance for the same agreement.
